stan/math/rev/mat/fun/cov_periodic.hpp

```diff
@@ -0,0 +1,380 @@
+     * input \f$ \mathbf{X} \f$ are defined as
+     * \f$ \mathbf{K}_{ij} = k(\mathbf{X}_i,\mathbf{X}_j), \f$ where
+     * \f$ \mathbf{X}_i \f$ is the \f$i\f$-th row of \f$ \mathbf{X} \f$ and \n
+     * \f$ k(\mathbf{x},\mathbf{x}^\prime) =
+     * \sigma^2 \exp\left(-\frac{2\sin^2(\pi |\mathbf{x}-\mathbf{x}^\prime|/p)}{\ell^2}\right), \f$
+     * \n
+     * where \f$ \sigma^2 \f$, \f$ \ell \f$ and \f$ p \f$ are the signal variance, length-scale and period.
+     *
+     * The partial derivatives w.r.t. the parameters are the following:\n
+     *
+     * \f$ \frac{\partial k}{\partial \sigma} = \frac{2k}{\sigma} \f$\n
+     * \f$ \frac{\partial k}{\partial \ell} = \frac{4k}{\ell^3} \sin^2(\pi|\mathbf{x}-\mathbf{x}^\prime|/p) \f$\n
+     * \f$ \frac{\partial k}{\partial p} = \frac{2k\pi|\mathbf{x}-\mathbf{x}^\prime|}{\ell^2p^2} \sin(2\pi|\mathbf{x}-\mathbf{x}^\prime|/p) \f$\n
+     * @tparam T_x type of std::vector elements of x
+     *   T_x can be a scalar, an Eigen::Vector, or an Eigen::RowVector.
+     * @tparam T_l type of length-scale
+     * @tparam T_p type of period
+     */
+    template <typename T_x, typename T_l, typename T_p>
+    class cov_periodic_vari<T_x, double, T_l, T_p> : public vari {
+    public:
+      const size_t size_;
+      const size_t size_ltri_;
+      const double l_d_;
+      const double sigma_d_;
+      const double p_d_;
+      const double sigma_sq_d_;
+      double* dist_;
+      double* sin_dist_;
+      double* cos_dist_;
+      double* sin_dist_sq_;
+      vari* l_vari_;
+      vari* p_vari_;
+      vari** cov_lower_;
+      vari** cov_diag_;
+
+      /**
+       * Constructor for cov_periodic.
+       *
+       * All memory allocated in
+       * ChainableStack's stack_alloc arena.
+       *
+       * It is critical for the efficiency of this object
+       * that the constructor create new varis that aren't
+       * popped onto the var_stack_, but rather are
+       * popped onto the var_nochain_stack_. This is
+       * controlled by the second argument to
+       * vari's constructor.
+       *
+       * @param x std::vector of input elements.
+       *   Assumes that all elements of x have the same size.
+       * @param sigma standard deviation of the signal
+       * @param l length-scale
+       * @param p period
+       */
+      cov_periodic_vari(const std::vector<T_x>& x, double sigma, const T_l& l,
+                        const T_p& p)
+        : vari(0.0),
+          size_(x.size()),
+          size_ltri_(size_ * (size_ - 1) / 2),
+          l_d_(value_of(l)), sigma_d_(sigma),
+          p_d_(value_of(p)),
+          sigma_sq_d_(sigma_d_ * sigma_d_),
+          dist_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
+          sin_dist_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
+          cos_dist_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
+          sin_dist_sq_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
```

Comment:
Some quick timing checks showed that there is around 16% performance gain when the values are stored rather than calculated. Do you consider this significant?

Reply:
Now removed cos_dist_ as calculating sin(2*x) instead of sin(x)*cos(x) is more optimal